src/Compilers/CSharp/Portable/FlowAnalysis/PreciseAbstractFlowPass.cs

```diff
@@ -2733,6 +2733,41 @@ public override BoundNode VisitDiscardExpression(BoundDiscardExpression node)
         {
             return null;
         }
+
+        public override BoundNode VisitConstructorMethodBody(BoundConstructorMethodBody node)
+        {
+            Visit(node.Initializer);
+            VisitMethodBodies(node.BlockBody, node.ExpressionBody);
+            return null;
+        }
+
+        public override BoundNode VisitNonConstructorMethodBody(BoundNonConstructorMethodBody node)
+        {
+            VisitMethodBodies(node.BlockBody, node.ExpressionBody);
+            return null;
+        }
+
+        private void VisitMethodBodies(BoundBlock blockBody, BoundBlock expressionBody)
+        { 
+            if (blockBody == null)
+            {
+                Visit(expressionBody);
+                return;
+            }
+            else if (expressionBody == null)
+            {
+                Visit(blockBody);
+                return;
+            }
+
+            LocalState initialState = this.State.Clone();
+            Visit(blockBody);
+            LocalState afterBlock = this.State;
+            SetState(initialState);
+            Visit(expressionBody);
```

Comment:
This feels overengineered. I think you could just as well visit one followed by the other (in syntax order).

Reply:
> This feels overengineered. I think you could just as well visit one followed by the other (in syntax order).

I think this is the right way to visit this node in the long run. I am keeping the implementation, unless you can come up with a test scenario that would demonstrate that the way that you are proposing gives correct result and the one that is implemented give incorrect result.

---
In reply to: [170053230](https://github.com/dotnet/roslyn/pull/24918#discussion_r170053230) [](ancestors = 170053230)

Reply:
We do not have a specification for flow analysis in the face of syntax errors, so there is no meaningful concept of a scenario that demonstrates which results are "correct". However, the usual meaning of one piece of code followed by another is that they are executed in sequence. You are implementing the semantics of some kind of conditional execution, but without any suggestion as to why that would be appropriate.

Reply:
>  is that they are executed in sequence

I view it as these are two unrelated pieces of code, they are definitely not executed one after another. As a feature owner, I am making a call, I will add a comment to explain the rational.

---
In reply to: [170117730](https://github.com/dotnet/roslyn/pull/24918#discussion_r170117730) [](ancestors = 170117730)